### PR TITLE
[RHCLOUD-31068] add Service Account support

### PR DESCRIPTION
**Pull request by g-duval**

### CODE REVIEW

File: src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        return serviceAccount;
+    }
+
+    @Override
+    public String getSubject() {
+        return getServiceAccount().getUsername();
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class ServiceAccount {
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java:26]: Should we check the scope of the service account? Maybe we should reject the request if that scope is not `api.notifications`. I'm not sure.

File: src/test/java/com/redhat/cloud/notifications/IdentityTest.java [hunk 2]
```diff
@@ -96,4 +96,37 @@ void testX509Identity() {
...
+                "    \"cross_access\" : false\n" +
+                "  }\n" +
+                "}" +
+            "}";
+
+        String xRhEncoded = null;
+        try {
+            xRhEncoded = new String(Base64.getEncoder().encode(header.getBytes("UTF-8")));
+        } catch (UnsupportedEncodingException e) {
+            Assertions.fail();
```
Comment (gwenneg) [inline src/test/java/com/redhat/cloud/notifications/IdentityTest.java:123]: This hides the exception in the test logs. We should add `throws Exception` to the test method instead.

File: src/main/java/com/redhat/cloud/notifications/GwResource.java [hunk 3]
```diff
@@ -285,14 +287,14 @@ private static String buildResponseEntity(boolean success, String details) {
...
     }
 
-    private Optional<X509Certificate> getX509Certificate(String bundle, String app, String certificateDn) {
+    private Optional<SourceEnvironment> getSourceEnvironment(String bundle, String app, String identificationData) {
         try {
-            X509Certificate x509Certificate = restValidationClient.validateCertificate(bundle, app, certificateDn);
-            Log.infof("Certificate validated, coming from source environment %s", x509Certificate.sourceEnvironment);
-            return Optional.of(x509Certificate);
+            SourceEnvironment sourceEnvironment = restValidationClient.validateCertificate(bundle, app, identificationData);
+            Log.infof("Identification validated, coming from source environment %s", sourceEnvironment.name);
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/notifications/GwResource.java:293]: General naming suggestion in this method:
```suggestion
            Log.infof("Authentication validated, coming from source environment %s", sourceEnvironment.name);
```

File: src/main/java/com/redhat/cloud/notifications/RestValidationClient.java [hunk 4]
```diff
@@ -34,6 +34,6 @@ public interface RestValidationClient {
     @Consumes(APPLICATION_JSON)
     @Produces(APPLICATION_JSON)
     @CacheResult(cacheName = "certificate-validation")
-    X509Certificate validateCertificate(@RestQuery String bundle, @RestQuery String application, @RestQuery String certificateSubjectDn);
+    SourceEnvironment validateCertificate(@RestQuery String bundle, @RestQuery String application, @RestQuery String certificateSubjectDn);
```
[the remark this answers is not in the document]
Comment (g-duval) [reply src/main/java/com/redhat/cloud/notifications/RestValidationClient.java:37]: I agree, should be done later.

File: src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java (hunk 1 above)
Comment (g-duval) [reply src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java:26]: I'm not sure if the scope matters here, let see how turnpike will be updated to restrict access to some routes, like it does with X509.

File: src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/notifications/auth/RhServiceAccountIdentity.java:26]:
```suggestion
    @JsonNaming(SnakeCaseStrategy.class)
    public static class ServiceAccount {
```

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**